stl/inc/ranges

```diff
@@ -22,15 +25,461 @@ _STL_DISABLE_CLANG_WARNINGS
 
 _STD_BEGIN
 namespace ranges {
-    // Much machinery defined in <xutility>
+    // MUCH machinery defined in <xutility>
 
     // clang-format off
     // CONCEPT ranges::viewable_range
     template <class _Rng>
     concept viewable_range = range<_Rng>
         && (borrowed_range<_Rng> || view<remove_cvref_t<_Rng>>);
+
+    // CLASS TEMPLATE ranges::_Semiregular_box
+    // clang-format off
+    template <class _Ty>
+    concept _Copy_constructible_object = copy_constructible<_Ty> && is_object_v<_Ty>;
+    // clang-format on
+
+#if 0 // TRANSITION, VSO-1174090
+    template <_Copy_constructible_object _Ty>
+    class _Semiregular_box {
+    public:
+        constexpr _Semiregular_box() noexcept : _Dummy{}, _Engaged{false} {}
+        constexpr _Semiregular_box() noexcept(is_nothrow_default_constructible_v<_Ty>) requires constructible_from<_Ty>
+            : _Val(), _Engaged{true} {}
+
+        template <class... _Types>
+        constexpr _Semiregular_box(in_place_t, _Types&&... _Args) noexcept(
+            is_nothrow_constructible_v<_Ty, _Types...>) // strengthened
+            : _Val(_STD forward<_Types>(_Args)...), _Engaged{true} {}
+
+        ~_Semiregular_box() requires is_trivially_destructible_v<_Ty> = default;
+
+        ~_Semiregular_box() {
+            if (_Engaged) {
+                _Val.~_Ty();
+            }
+        }
+
+        _Semiregular_box(const _Semiregular_box&) requires is_trivially_copy_constructible_v<_Ty> = default;
+        _Semiregular_box(const _Semiregular_box& _That) : _Engaged{_That._Engaged} {
+            if (_Engaged) {
+                _Construct_in_place(_Val, _That._Val);
+            }
+        }
+
+        _Semiregular_box(_Semiregular_box&&) requires is_trivially_move_constructible_v<_Ty> = default;
+        _Semiregular_box(_Semiregular_box&& _That) : _Engaged{_That._Engaged} {
+            if (_Engaged) {
+                _Construct_in_place(_Val, _STD move(_That._Val));
+            }
+        }
+
+        // clang-format off
+        _Semiregular_box& operator=(const _Semiregular_box&) noexcept
+            requires copyable<_Ty> && is_trivially_copy_assignable_v<_Ty> = default;
+        // clang-format on
+
+        _Semiregular_box& operator=(const _Semiregular_box& _That) noexcept(is_nothrow_copy_constructible_v<_Ty>&&
+                is_nothrow_copy_assignable_v<_Ty>) /* strengthened */ requires copyable<_Ty> {

```

Comment:
The spec speaks about the assignment operator if `_Ty` does not model [assignable_from](http://eel.is/c++draft/range.semi.wrap#1.3)

So I am wondering whether this is is strictly equivalent, given that `copyable` is a stronger constraint than `assignable_­from<T&, const T&>`

Reply:
The marginal constraint between `copyable<T>` and `copy_constructible<T>` is not precisely equivalent to `assignable_from<T&, const T&>`, although it was intended to be so when this wording was written. (Sometime between then and finishing up C++20 we changed the wording for implicit expression variations in [concepts.equality], which broke that equivalence.) I'll file an LWG issue.

Reply:
Issue submitted:

> The exposition-only semiregular-box class template specified in [range.semi.wrap] implements a default constructor, copy assignment operator, and move assignment operator atop the facilities provided by std::optional when the wrapped type is not default constructible, copy assignable, or move assignable (respectively). The constraints on the copy and move assignment operator implementations go out of their way to be unnecessarily minimal. The meaning of the constraint on the copy assignment operator - !assignable<T, const T&> - has even changed since this wording was written as a result of LWG reformulating the implicit expression variations wording in [concepts.equality].
>
> It would be much simpler for implementors and users if we recall that minimality is not the primary goal of constraints and instead constrain these assignment operators more simply with !movable<T> and !copyable<T>.
>
> ### Proposed Resolution
> Wording relative to N4861.
>
> Change [range.semi.wrap] as follows:
>
> ```diff
>  (1.3) — If
> -assignable_from<T&, const T&>
> +copyable<T>
>  is not modeled, the copy assignment operator is equivalent to:
> 
>  [...]
>  
>  (1.4) — If 
> -assignable_from<T&, T>
> +movable<T>
>  is not modeled, the move assignment operator is equivalent to:
>  
>  [...]
> ```